### DISCUSSION TITLE
Specifying Build-Dependency on Generated Headers

### DISCUSSION THREAD

**Discussion by lhearachel** [mesonbuild/meson#13107] [Q&A] [answered]
I am maintaining a project where we compile an incoming data file of string-rows into a binary format using a custom encoder program. As part of that process, our translation layer has an option to output a generated header file for inclusion in the C source code. We've included the directory in which these headers are generated to our project's `include_directories`, but I'm unconvinced that this is not brittle, even though it seems to work consistently.

Is there a best practice beyond this for ensuring that this process always runs before any source-code files are passed to compilation? Or does meson automagically trace that dependency from the `#include` directives inside the source-code files?

**Comment by eli-schwartz** [comment, marked as answer]
> Is there a best practice beyond this for ensuring that this process always runs before any source-code files are passed to compilation? Or does meson automagically trace that dependency from the `#include` directives inside the source-code files?

Both.

- You need to add the custom_target output as one of the source files for any target that has C files which includes the generated header. Otherwise, depending on scheduling, the C file might try to compile before the header has been created for the first time, and it will fail. This is only relevant for the first time building, as it's a soft dependency. It is impossible to bootstrap the dependency tracker until compilation runs at least once, so the header has to exist (even if it is out of date) before trying to compile the sources.
- meson will automagically track `#include` directives using compiler "depfile" support, which ensures *rebuild* correctness: every time the header is regenerated, the C sources using it are marked out of date.


If a target has multiple C files (only one of which uses the header) and also the generated header, the generated header must be built for the first time before *any* C sources are compiled. After that, regenerating the header will only rebuild the files that the depfile scanner actually reported as using the header.

**Reply by dcbaker** [reply]
We have a fairly thorough explanation of what needs to happen here: https://mesonbuild.com/FAQ.html#how-do-i-tell-meson-that-my-sources-use-generated-headers
